src/common/fair_mutex.h

```diff
@@ -2,22 +2,48 @@
 
 #pragma once
 
-#include "common/ceph_mutex.h"
+#include "include/ceph_assert.h"
 
 #include <thread>
 #include <string>
+#include <mutex>
+#include <condition_variable>
 
 namespace ceph {
 /// a FIFO mutex
 class fair_mutex {
 public:
-  fair_mutex(const std::string& name)
-    : mutex{ceph::make_mutex(name)}
-  {}
+  fair_mutex(const std::string& name) {}
   ~fair_mutex() = default;
   fair_mutex(const fair_mutex&) = delete;
   fair_mutex& operator=(const fair_mutex&) = delete;
 
+#ifdef CEPH_DEBUG_MUTEX
+  pthread_mutex_t* native_handle() {
+    return mutex.native_handle();
+  }
+
+  void _pre_unlock() {
+    mutex.lock();
+    ceph_assert(locked_by == std::this_thread::get_id());
+    ++unblock_id;
+    locked_by = std::thread::id();
+    // wake up all the waiters before giving up the lock
+    // in condition_variable_debug's wait()
+    int r = pthread_cond_broadcast(cond.native_handle());
```

Comment:
i'd prefer not do unlock in `__pre_unlock()`. it's just difficult to reason about.

IMHO, `++unblock_id` is part of unlock of fair_mutex. so i'd prefer keeping `fair_mutex` simple.

Reply:
We just need to wake up all the other waiters, who are stuck in `fair_mutex::lock()`, before calling the `pthread_cond_wait()` later in `condition_variable_debug::wait()`.

Reply:
Anyway this will be removed for now in this PR.